tracing/src/test/java/org/axonframework/extensions/tracing/TracingQueryGatewayTest.java

```diff
@@ -43,29 +53,113 @@ void before() {
                                          .delegateQueryBus(mockQueryBus)
                                          .build();
 
-        answer = new GenericQueryResponseMessage<>("answer");
+        answer1 = new GenericQueryResponseMessage<>("answer1");
+        answer2 = new GenericQueryResponseMessage<>("answer2");
     }
 
     @Test
-    void testQuery() throws ExecutionException, InterruptedException {
+    void testQuery_queryName() throws ExecutionException, InterruptedException {
         //noinspection unchecked
         when(mockQueryBus.query(any(QueryMessage.class)))
-                .thenReturn(CompletableFuture.completedFuture(answer));
+                .thenReturn(CompletableFuture.completedFuture(answer1));
 
         MockSpan span = mockTracer.buildSpan("test").start();
         ScopeManager scopeManager = mockTracer.scopeManager();
         try (final Scope ignored = scopeManager.activate(span)) {
-            CompletableFuture<String> query = testSubject.query("query", "Query", String.class);
-            assertEquals("answer", query.get());
+            CompletableFuture<String> query = testSubject.query("pointQuery", "Query", String.class);
+            assertEquals("answer1", query.get());
 
             // Verify the parent span is restored, and that a child span was finished.
             Span activeSpan = mockTracer.activeSpan();
             assertEquals(span, activeSpan);
 
             List<MockSpan> mockSpans = mockTracer.finishedSpans();
             assertEquals(1, mockSpans.size());
-            assertEquals("send_query", mockSpans.get(0).operationName());
+            assertEquals("query_pointQuery", mockSpans.get(0).operationName());
         }
         assertNull(scopeManager.activeSpan(), "There should be no activeSpan");
     }
+
+    @Test
+    void testQuery_query() throws ExecutionException, InterruptedException {
+        //noinspection unchecked
+        when(mockQueryBus.query(any(QueryMessage.class)))
+                .thenReturn(CompletableFuture.completedFuture(answer1));
+
+        MockSpan span = mockTracer.buildSpan("test").start();
+        ScopeManager scopeManager = mockTracer.scopeManager();
+        try (final Scope ignored = scopeManager.activate(span)) {
+            CompletableFuture<String> query = testSubject.query(
+                    new GenericMessage<>("Query"),
+                    String.class);
+            assertEquals("answer1", query.get());
+
+            // Verify the parent span is restored, and that a child span was finished.
+            Span activeSpan = mockTracer.activeSpan();
+            assertEquals(span, activeSpan);
+
+            List<MockSpan> mockSpans = mockTracer.finishedSpans();
+            assertEquals(1, mockSpans.size());
+            assertEquals("query_GenericMessage", mockSpans.get(0).operationName());
```

Comment:
There's a slight discrepancy with the operation name here. The `QueryGateway` should return the `queryName` to be the name of the payload in the `GenericMessage`, not the class name of the `GenericMessage`. This requires a change in the `QueryGateway` though, hence something we cannot wait for directly.

By the way, I've adjusted the behaviour in the `QueryGateway` in [this](https://github.com/AxonFramework/AxonFramework/pull/1410) pull request.

Reply:
Thanks for that!